### PR TITLE
Dispatcher.set_commands

### LINKED ISSUE #1859: [FEATURE] Infere bot commands from CommandHandler and set them automatically
#### Is your feature request related to a problem? Please describe.
Even with API 4.7s new `Bot.set_my_commands`, setting commands will be tiresome

#### Describe the solution you'd like
* Add `description` attribute to `CommandHandler`
* Add `Disptacher.set_commands(update=False, delete=False, add=True)` that
  * gets the bots command from the registered `CommandHandlers`
  * Checks that they are valid for `BotCommand` and `set_my_commands`
  * gets current commands
  * sets new commands with options:
    * `update`: If `False` don't change a command if already set
    * `delete`: If `True`, delete commands that were set but currently don't have a corresponding handler
    * `add`: If `False` and currently commands are set, doesn't add new ones

#### Describe alternatives you've considered
Tiresome handwork

#### Additional context
`Bot.set_my_commands` is very new, so maybe wait till the next API update to make sure Telegram doesn't change the methods directly

### PR DESCRIPTION
**Pull request by Bibo-Joshi**
Proposal implementation for a feature as discussed in #1856 
Tests can be simplified, once #1724 is merged.

Closes #1859

### CODE REVIEW

**Comment by lucasmezencio** [comment]
Hey @Bibo-Joshi could you fix the conflicts? Let's try to push this to be merged. 🙂

File: telegram/ext/commandhandler.py [hunk 1]
```diff
@@ -124,14 +128,19 @@ def __init__(self,
...
         super(CommandHandler, self).__init__(
             callback,
             pass_update_queue=pass_update_queue,
             pass_job_queue=pass_job_queue,
             pass_user_data=pass_user_data,
             pass_chat_data=pass_chat_data)
 
+        self.description = description
+        if self.description is not None and not 3 <= len(self.description) <= 256:
+            raise ValueError('Command description is not valid.')
```
Comment (Bibo-Joshi) [inline telegram/ext/commandhandler.py:142]: I should remove this. We don't usually perform such checks as TG will just return BadRequest if invalid and more importantly, the limitations might change

File: telegram/ext/dispatcher.py [hunk 2]
```diff
@@ -529,3 +529,72 @@ def dispatch_error(self, update, error):
...
+            add (:obj:`bool`, optional): If there are commands set for the bot and :attr:`add` is
+                :obj:`False`, no new commands are added. Defaults to :obj:`True`.
+            update (:obj:`bool`, optional): Whether to override descriptions of commands already
+                set. Defaults to :obj:`True`.
+            delete (:obj:`bool`, optional): Whether to delete commands, which are currently set but
+                don't have a corresponding :class:`telegram.ext.CommandHandler` listening for that
+                command. Defaults to :obj:`True`.
+            skip_empty (:obj:`bool`, optional): Whether to skip
+                :class:`telegram.ext.CommandHandler` s with an empty description. If :obj:`False`,
+                empty descriptions will be replaced by ``Command "<command>"``.
```
Comment (Bibo-Joshi) [inline telegram/ext/dispatcher.py:556]: Defaults to :obj:`False`

File: telegram/ext/dispatcher.py [hunk 3]
```diff
@@ -529,3 +529,72 @@ def dispatch_error(self, update, error):
...
+            alphabetical (:obj:`bool`, optional): Whether to sort commands by alphabetical order.
+                If :obj:`False`, commands are sorted in the same order, in with :meth:`add_handler`
```
Comment (Bibo-Joshi) [inline telegram/ext/dispatcher.py:558]: with -> which

File: telegram/ext/dispatcher.py [hunk 4]
```diff
@@ -529,3 +529,72 @@ def dispatch_error(self, update, error):
...
+                is invoked. If :attr:`delete` is :obj:`False`, commands without corresponding
+                handler will be last. Defaults to :obj:`False`.
+        """
+        set_bot_commands = self.bot.get_my_commands()
+
+        dp_bot_commands = []
+        for group in self.handlers:
+            for handler in self.handlers[group]:
+                if hasattr(handler, 'command') and hasattr(handler, 'description'):
```
Comment (Bibo-Joshi) [inline telegram/ext/dispatcher.py:567]: Rather just do `isinstance(handler, CommandHandler)`. Also we should check `ConversationHandlers` for `CommandHandlers`

**Comment by Bibo-Joshi** [comment]
By now I'm not really confident anymore, that this is a good addition. It doesn't really save too many code lines and it probably covers only some use cases. Maybe it's better suited for an eventual contrib repo, see #1912. I'll remove from the v13 milestone for now

**Comment by JosXa** [comment]
tl;dr Reasoning?

**Comment by Bibo-Joshi** [comment]
> tl;dr Reasoning?

[this](https://github.com/python-telegram-bot/python-telegram-bot/pull/1911#issuecomment-671053598)

**Comment by JosXa** [comment]
> > tl;dr Reasoning?
> 
> [this](https://github.com/python-telegram-bot/python-telegram-bot/pull/1911#issuecomment-671053598)

I only see

> It doesn't really save too many code lines and it probably covers only some use cases

but I would have merged that :P

**Comment by Bibo-Joshi** [comment]
My reasoning is basically the same as why #854 got closed: There are valid use cases for this feature, but it can only ever cover a subset of the use cases. So we can merge and find ourselfes adding more and more as users want more use cases - or we don't and save the work of maintaining a very specific feature. And at least currently I prefere the latter … Mind you, I'm still very open for #1912! :)
